torch/distributed/_shard/sharded_tensor/api.py

```diff
@@ -273,46 +278,85 @@ def gather(
+        rank_sizes = [0 for _ in range(world_size)]
+        max_rank_size = 0
+        shard_placement: Dict[int, Tuple[int, int]] = dict()
+        local_shards_placement = []
+        # collect sizes
+        for shard_idx, shard_md in enumerate(self.metadata().shards_metadata):
+            shard_rank = cast(_remote_device, shard_md.placement).rank()
+            assert shard_rank is not None
+            shard_placement[shard_idx] = (shard_rank, rank_sizes[shard_rank])
+            if shard_rank == rank:
+                local_shards_placement.append((shard_md, rank_sizes[shard_rank],))
+
+            rank_sizes[shard_rank] += shard_size(shard_md)
+            max_rank_size = max(max_rank_size, rank_sizes[shard_rank])
+
+        gather_list: Optional[List[torch.Tensor]]
+        if rank == dst:
+            assert out is not None
+            gather_list = [torch.empty((max_rank_size,), device=out.device) for _ in range(world_size)]
+        else:
+            gather_list = None
+
+        with torch.no_grad():
+            # This is an aproximation for what device to use.
+            # The lack of policy enforcement around is pretty bad
+            device: Union[torch.device, str]
+            if len(local_shards) > 0:
+                device = local_shards[0].tensor.device
+            else:
+                device = f"cuda:{torch.cuda.current_device()}"
+            data = torch.empty(max_rank_size, device=device)
+            for shard in local_shards:
+                for placement in local_shards_placement:
+                    if placement[0] == shard.metadata:
+                        src = shard.tensor.flatten()
+                        data[placement[1]: placement[1] + src.numel()].copy_(src)
+                        break
+
+        dist.gather(
+            tensor=data,
+            gather_list=gather_list,
             dst=dst,
             group=self._process_group,
         )
-        if rank == dst:
-            if out is None:
-                raise ValueError("`out` Tensor must be provided on dst rank!")
-            dims = len(full_size)
-            for shards in gathered_shards:
-                if shards is None:
-                    raise RuntimeError(
-                        'Gathered shards cannot be None on dst rank {dst}'
-                    )
-                for shard in shards:
-                    metadata = shard.metadata
-                    tensor = shard.tensor
-
-                    out_narrow_view = out
-                    for dim in range(dims):
-                        out_narrow_view = out_narrow_view.narrow(
-                            dim,
-                            metadata.shard_offsets[dim],
-                            metadata.shard_sizes[dim],
-                        )
-
-                    out_narrow_view.copy_(tensor)
+        if rank != dst:
+            return
+        if out is None:
+            raise ValueError("`out` Tensor must be provided on dst rank!")
```

Comment:
Probably do this validation right at the top of the function, not after doing expensive collectives.

Reply:
This check is already done by _validate_output_tensor_for_gather but mypy can't see it. 
Replacing it with a cast and a comment on why.